### ISSUE TITLE
Known issues for VIRTUAL and STORED client side computed columns

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13685]
jOOQ 3.17 introduced a big feature called client side computed columns (https://github.com/jOOQ/jOOQ/issues/9879). There are a few open issues including:
### Tasks

The following changes are needed:

- [x] Code generation
  - [ ] The combination with `name`, `converter`, or `binding` must be tested as well
  - [ ] The combinations above that aren't yet supported should be logged
- [x] Runtime
  - [ ] Settings to enable / disable the feature at runtime
- [x] Tests
  - [ ] A `VIRTUAL` computed column that depends on a `STORED` computed column
  - [ ] `VIRTUAL` computed columns contained in embeddables
    - [ ] with `replacesFields = false`
    - [ ] with `replacesFields = true`
  - [ ] `STORED` computed columns
    - [ ] `INSERT` with aliased table
    - [ ] Reorder computed columns again to prevent interdependency side effects, if any (in MySQL)
    - [ ] `INSERT .. DEFAULT VALUES`
  - [ ] `UPDATE`
    - [ ] `UPDATE .. FROM`
  - [ ] `MERGE`
    - [ ] This is extremely hard to emulate correctly. I guess we can't ship support for the feature in `MERGE` statements for now, unless the `MERGE` statement is generated by jOOQ's internals (e.g. `ON CONFLICT` emulation)
  - [ ] Combine client side computed columns with readonly columns
  - [ ] Combine with `Settings.readonlyInsert` and other settings
  - [ ] Combine with embeddables
  - [ ] The SQL transformation should have a simple and sophisticated version
    - [ ] Simple is for `Param` types (i.e. constants) or deterministic expressions based on `Param` types ("effectively" constants), in case the query shouldn't be changed drastically
  - [ ] `STORED` computed columns contained in embeddables
    - [ ] with `replacesFields = false`
    - [ ] with `replacesFields = true`

### Related work / follow up work:

- https://github.com/jOOQ/jOOQ/issues/13411
- https://github.com/jOOQ/jOOQ/issues/13418

### Caveats:

- [ ] Some queries cannot enforce computed columns in all dialects, if any given syntax cannot be emulated. This requires
  - [ ] Sufficiently clear error handling
  - [ ] Documentation
  - [ ] Tests that the computation either works, or fails with an error. It's important we don't silently execute the query with a wrong result
- [x] MySQL `UPDATE` doesn't run the `SET` clause atomically, see https://twitter.com/lukaseder/status/1507019364800688134, https://stackoverflow.com/q/37649/521799. This has 2 implications:
  - [x] Computed columns must be calculated lexically after all user-provided `SET` clauses
  - [ ] We have to resolve the inter dependency tree of computational expressions to emulate standard atomic computation behaviour. Until that is available, the behaviour in case of interdependent expressions is undefined.
- [ ] `Field.getDataType()` currently produces the underlying data type from the schema, not the one from the computational expression, which may differ (e.g. have different nullability or a specific row type in case of `MULTISET` or `ROW`). We should make the correct type available statically.

**Comment by lukaseder** [comment]
Thanks, @mkurz. This is about client side computed columns. Why not create a new issue requesting support for this in PostgreSQL?
